composeApp/src/main/kotlin/io/github/couchtracker/ui/screens/viewing/DateTimeSection.kt

```diff
@@ -0,0 +1,332 @@
+import io.github.couchtracker.ui.screens.viewing.DateTimeSectionState.CustomDateDialogVisibility.TIMEZONE
+import io.github.couchtracker.utils.Text
+import io.github.couchtracker.utils.roundToSeconds
+import io.github.couchtracker.utils.str
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.toLocalDateTime
+import kotlin.time.Duration
+
+class DateTimeSectionState {
+    var dateTime by mutableStateOf<DateAndTimeValue?>(null)
+    var customDateDialogVisibility by mutableStateOf(HIDDEN)
+    var customTimezone by mutableStateOf<TimeZone?>(TimeZone.currentSystemDefault())
+
+    fun setCustomDate(
+        localDateTime: PartialDateTime.Local? = dateTime?.dateTime?.local,
+        timeZone: TimeZone? = customTimezone,
+    ) {
+        customTimezone = timeZone
+        dateTime = if (localDateTime == null) {
+            null
+        } else {
+            DateAndTimeValue(
+                DateAndTimeSectionChoices.CUSTOM,
+                if (timeZone == null) {
+                    localDateTime
+                } else {
+                    PartialDateTime.Zoned(localDateTime, timeZone)
+                },
+            )
+        }
+    }
+
+    enum class CustomDateDialogVisibility {
+        HIDDEN,
+        DATE,
+        TIME,
+        TIMEZONE,
+    }
+}
+
+data class DateAndTimeValue(
+    val category: DateAndTimeSectionChoices,
+    val dateTime: PartialDateTime,
+)
+
+enum class DateAndTimeSectionChoices(
+    val displayName: Text,
+    val default: (Instant, TimeZone, runtime: Duration) -> PartialDateTime,
+) {
+    JUST_STARTED(
+        Text.Resource(R.string.create_viewing_time_just_started),
+        { now, tz, runtime ->
+            PartialDateTime.Zoned(
+                local = PartialDateTime.Local.DateTime(now.toLocalDateTime(tz).roundToSeconds()),
+                zone = tz,
+            )
+        },
+    ),
+    JUST_FINISHED(
+        Text.Resource(R.string.create_viewing_time_just_finished),
+        { now, tz, runtime ->
+            PartialDateTime.Zoned(
+                local = PartialDateTime.Local.DateTime(now.minus(runtime).toLocalDateTime(tz).roundToSeconds()),
+                zone = tz,
+            )
+        },
+    ),
+    TODAY(
+        Text.Resource(R.string.create_viewing_time_today),
+        { now, tz, runtime ->
+            PartialDateTime.Zoned(
+                local = PartialDateTime.Local.Date(now.toLocalDateTime(tz).date),
+                zone = tz,
+            )
+        },
+    ),
+    CUSTOM(
+        Text.Resource(R.string.create_viewing_time_custom),
+        { _, _, _ -> throw UnsupportedOperationException() },
+    ),
+}
+
+@Composable
+fun rememberDateTimeSectionState(): DateTimeSectionState {
+    return remember {
+        DateTimeSectionState()
+    }
+}
+
+@Composable
+fun DateTimeSectionDialog(sectionState: DateTimeSectionState) {
```

Comment:
This has the same problem as above: shouldn't we pass immutable data to the composables (especially public ones) and have `value`/`onValueChange` parameters?